### PR TITLE
`Response`-related classes refactoring

### PR DESCRIPTION
**Pull request by vcolin7**
Fixes: #38075.

Changes in this PR:
1. [`Response<T>`](https://github.com/Azure/azure-sdk-for-java/blob/102857167dfe12bc5aa9047540ce7c1c92b1ab2b/sdk/generic-sdk-core/generic-core/src/main/java/com/generic/core/http/Response.java) now extends `Closeable` and has a new `getBody()` method that returns a `BinaryData`.
2. [`HttpResponse<T>`](https://github.com/Azure/azure-sdk-for-java/blob/102857167dfe12bc5aa9047540ce7c1c92b1ab2b/sdk/generic-sdk-core/generic-core/src/main/java/com/generic/core/http/models/HttpResponse.java) is now parameterized and the default implementation of `Response<T>` for generic-core. Other packages can still extend their own subtypes (e.g. [`OkHttpResponse`).](https://github.com/Azure/azure-sdk-for-java/blob/102857167dfe12bc5aa9047540ce7c1c92b1ab2b/sdk/generic-sdk-core/generic-core-http-okhttp/src/main/java/com/generic/core/http/okhttp/implementation/OkHttpResponse.java)
4. Virtually all internal APIs that accepted or returned `HttpResponse` now accept or return `Response<?>` instead. (e.g. `HttpClient.send()`, now returns a `Response<?>`).
5. `HttpResponseDecoder` has been removed but the response body is still deserialized in the same cases as before. The deserialized body is then set onto the original response as its `value`.
6. The internal `RestProxyBase.createResponse()` has been renamed to [`RestProxyBase.createResponseIfNecessary()`](https://github.com/Azure/azure-sdk-for-java/blob/102857167dfe12bc5aa9047540ce7c1c92b1ab2b/sdk/generic-sdk-core/generic-core/src/main/java/com/generic/core/implementation/http/rest/RestProxyBase.java#L95) and now simply returns the original response created by `HttpClient` if the return type of the `ServiceIntrface` operation is `Response` (we no longer use `Class.cast()`). Otherwise, we still try to create the appropriate `Response` subtype as usual.

For a general idea of how deserialization was originally handled in Generic Core, see [this gist](https://gist.github.com/vcolin7/9ce030a5913237560af4b96a09eea27c).

### CODE REVIEW

File: sdk/generic-sdk-core/generic-core-http-okhttp/src/main/java/com/generic/core/http/okhttp/implementation/OkHttpBufferedResponse.java [hunk 1]
```diff
@@ -10,18 +10,23 @@
 /**
  * An OkHttp response where the response body has been buffered into memory.
  */
-public final class OkHttpBufferedResponse extends OkHttpResponseBase {
-    private final byte[] body;
+public final class OkHttpBufferedResponse extends OkHttpResponseBase<byte[]> {
```
Comment (alzimmermsft) [inline sdk/generic-sdk-core/generic-core-http-okhttp/src/main/java/com/generic/core/http/okhttp/implementation/OkHttpBufferedResponse.java:13]: This change makes me think this should be changed to `OkHttpResponseBase<BinaryData>` instead of `byte[]`

File: sdk/generic-sdk-core/generic-core-http-okhttp/src/main/java/com/generic/core/http/okhttp/implementation/OkHttpResponse.java [hunk 2]
```diff
@@ -11,13 +11,15 @@
 /**
  * Default HTTP response for OkHttp.
  */
-public final class OkHttpResponse extends OkHttpResponseBase {
+public final class OkHttpResponse extends OkHttpResponseBase<byte[]> {
```
Comment (alzimmermsft) [inline sdk/generic-sdk-core/generic-core-http-okhttp/src/main/java/com/generic/core/http/okhttp/implementation/OkHttpResponse.java:14]: With our usage of `BinaryData` everywhere I think `OkHttpBufferedResponse` and `OkHttpResponse` can be merged and the appropriate buffering handling and `getValue` can be determined by `BinaryData`'s APIs

File: sdk/generic-sdk-core/generic-core/src/main/java/com/generic/core/implementation/http/serializer/HttpResponseDecoder.java [hunk 3]
```diff
@@ -32,22 +33,18 @@ public HttpResponseDecoder(ObjectSerializer serializer) {
...
     /**
-     * A decorated HTTP response which has subscribable body and headers that supports lazy decoding.
-     *
-     * <p>Subscribing to body kickoff http content reading, it's decoding then emission of decoded object. Subscribing
-     * to header kickoff header decoding and emission of decoded object.
+     * A decorated HTTP response that supports lazy decoding.
      */
-    public static class HttpDecodedResponse implements Closeable {
-        private final HttpResponse response;
+    public static class HttpDecodedResponse extends HttpResponse<Object> {
```
Comment (alzimmermsft) [inline sdk/generic-sdk-core/generic-core/src/main/java/com/generic/core/implementation/http/serializer/HttpResponseDecoder.java:43]: Is there a reason that this implements `HttpResponse` now? I'm not against it, just don't fully understand the change.

Comment (vcolin7) [reply sdk/generic-sdk-core/generic-core/src/main/java/com/generic/core/implementation/http/serializer/HttpResponseDecoder.java:43]: It felt better to treat `HttpDecodedResponse` as a type of `HttpResponse` than to have the former store the latter as a property and use `getSourceResponse()` to access it every time we needed to check a request, status code, or any of its other properties.

Conceptually I also like the idea of having it be a type of `HttpResponse`.

File: sdk/generic-sdk-core/generic-core/src/main/java/com/generic/core/http/policy/RedirectPolicy.java [hunk 4]
```diff
@@ -40,15 +40,19 @@ public RedirectPolicy(RedirectStrategy redirectStrategy) {
     @Override
     public HttpResponse process(HttpRequest httpRequest, HttpPipelineNextPolicy next) {
         // Reset the attemptedRedirectUrls for each individual request.
-        return attemptRedirect(httpRequest, next, 1, new HashSet<>());
+        try {
+            return attemptRedirect(httpRequest, next, 1, new HashSet<>());
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
```
Comment (samvaity) [inline sdk/generic-sdk-core/generic-core/src/main/java/com/generic/core/http/policy/RedirectPolicy.java:46]: throw LOGGER.logThrowableAsError(new UncheckedIOException(e));

File: sdk/generic-sdk-core/generic-core/src/main/java/com/generic/core/http/policy/RedirectPolicy.java [hunk 5]
```diff
@@ -40,15 +40,19 @@ public RedirectPolicy(RedirectStrategy redirectStrategy) {
...
+        }
     }
 
     /**
      * Function to process through the HTTP Response received in the pipeline and redirect sending the request with a
      * new redirect URL.
      */
     private HttpResponse attemptRedirect(final HttpRequest httpRequest, final HttpPipelineNextPolicy next,
-                                         final int redirectAttempt, Set<String> attemptedRedirectUrls) {
+                                         final int redirectAttempt, Set<String> attemptedRedirectUrls) throws IOException {
```
Comment (samvaity) [inline sdk/generic-sdk-core/generic-core/src/main/java/com/generic/core/http/policy/RedirectPolicy.java:55]: why are these throwing IOException now?

File: sdk/generic-sdk-core/generic-core/src/main/java/com/generic/core/implementation/http/rest/RestProxyBase.java [hunk 6]
```diff
@@ -94,51 +92,49 @@ protected abstract Object invoke(Object proxy, Method method, RequestOptions opt
...
+        // Inspection of the response type needs to be performed to determine the course of action: either cast the
+        // decoded response to Response or rely on reflection to create an appropriate Response subtype.
+        if (clazz.equals(Response.class)) {
+            if (decodedResponse.getValue() == null) {
+                decodedResponse.setCachedBody(bodyAsObject);
+            }
 
-        return RESPONSE_CONSTRUCTORS_CACHE.invoke(constructorReflectiveInvoker, response, bodyAsObject);
+            // Return the received decoded response cast to Response.
+            return clazz.cast(decodedResponse);
```
Comment (samvaity) [inline sdk/generic-sdk-core/generic-core/src/main/java/com/generic/core/implementation/http/rest/RestProxyBase.java:108]: Wouldnt this throw ClassCastException as its attempting to cast to something that doesn't extend from Response.class?

Comment (vcolin7) [reply sdk/generic-sdk-core/generic-core/src/main/java/com/generic/core/implementation/http/rest/RestProxyBase.java]: It won't, `decodedResponse` is a subtype of `Response`.

File: sdk/generic-sdk-core/generic-core/src/main/java/com/generic/core/http/policy/RedirectPolicy.java (hunk 5 above)
Comment (vcolin7) [reply sdk/generic-sdk-core/generic-core/src/main/java/com/generic/core/http/policy/RedirectPolicy.java]: It shouldn't. I think I missed it when merging `main` into this branch.

File: sdk/generic-sdk-core/generic-core-http-okhttp/src/main/java/com/generic/core/http/okhttp/OkHttpHttpClient.java [hunk 7]
```diff
@@ -156,18 +155,18 @@ private static HttpResponse toHttpResponse(HttpRequest request, okhttp3.Response
...
+            return new OkHttpBufferedResponse(response, request, eagerlyConvertHeaders, EMPTY_BODY);
         }*/
 
         // Use a buffered response when we are eagerly reading the response from the network and the body isn't empty.
         if (eagerlyReadResponse || ignoreResponseBody) {
             try (ResponseBody body = response.body()) {
                 byte[] bytes = (body != null) ? body.bytes() : EMPTY_BODY;
 
-                return new OkHttpBufferedResponse(response, request, bytes, eagerlyConvertHeaders);
+                return new OkHttpResponse(response, request, eagerlyConvertHeaders, bytes);
```
[the remark this answers is not in the document]
Comment (vcolin7) [reply sdk/generic-sdk-core/generic-core-http-okhttp/src/main/java/com/generic/core/http/okhttp/OkHttpHttpClient.java:165]: We merged both OkHttp response types and handle buffering internally when necessary.

File: sdk/generic-sdk-core/generic-core/src/main/java/com/generic/core/http/policy/RedirectPolicy.java (hunk 5 above)
Comment (vcolin7) [reply sdk/generic-sdk-core/generic-core/src/main/java/com/generic/core/http/policy/RedirectPolicy.java]: I'll remove it.

File: sdk/generic-sdk-core/generic-core-http-okhttp/src/main/java/com/generic/core/http/okhttp/OkHttpHttpClient.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (vcolin7) [reply sdk/generic-sdk-core/generic-core-http-okhttp/src/main/java/com/generic/core/http/okhttp/OkHttpHttpClient.java:165]: To be clearer: the response itself does not know whether to buffer or not, we create the instance with the buffered body bytes in a `BinaryData` if either `eagerlyReadResponse` or `ignoreResponseBody` are set to `true`, like what we do on Azure Core. Otherwise, the `BinaryData` holds an `InputStream` we get from the OkHttp `ResponseBody`.

File: sdk/generic-sdk-core/generic-core/src/main/java/com/generic/core/implementation/http/rest/RestProxyBase.java (hunk 6 above)
Comment (samvaity) [reply sdk/generic-sdk-core/generic-core/src/main/java/com/generic/core/implementation/http/rest/RestProxyBase.java]: Can we add a test case for this?

File: sdk/generic-sdk-core/generic-core/src/samples/java/com/generic/core/http/annotation/AnnotationJavadocCodesnippets.java [hunk 8]
```diff
@@ -27,7 +27,8 @@ interface BodyParamExamples {
             + "{resourceGroupName}/providers/Microsoft.Compute/virtualMachines/{vmName}",
```
Comment (samvaity) [inline sdk/generic-sdk-core/generic-core/src/samples/java/com/generic/core/http/annotation/AnnotationJavadocCodesnippets.java:27]: Revert changes unrelated to PR.

File: sdk/generic-sdk-core/generic-core/src/samples/java/com/generic/core/util/ClientLoggerJavaDocCodeSnippets.java [hunk 9]
```diff
@@ -109,7 +106,7 @@ public void loggingSnippets() {
         }
         // END: com.generic.core.util.logging.clientlogger.atError#deffered-value
 
-        Response<Void> response = getResponse();
+        Response<?> response = Response.create(null, 200, null, null);
```
Comment (samvaity) [inline sdk/generic-sdk-core/generic-core/src/samples/java/com/generic/core/util/ClientLoggerJavaDocCodeSnippets.java:109]:
```suggestion
        Response<Void> response = Response.create(null, 200, null, null);
```

File: sdk/generic-sdk-core/generic-core/src/main/java/com/generic/core/http/models/HttpResponse.java [hunk 10]
```diff
@@ -3,90 +3,145 @@
...
      * @param request The {@link HttpRequest} that resulted in this {@link HttpResponse}.
-     * @param bodyBytes The response body as a byte array.
+     * @param statusCode The response status code.
+     * @param headers The response {@link Headers}.
+     * @param value The response body.
+     * @param isRawResponse Indicates whether the value provided needs to be deserialized.
      */
-    protected HttpResponse(HttpRequest request, byte[] bodyBytes) {
+    @SuppressWarnings("unchecked")
+    public HttpResponse(HttpRequest request, int statusCode, Headers headers, BinaryData value, boolean isRawResponse) {
```
Comment (samvaity) [inline sdk/generic-sdk-core/generic-core/src/main/java/com/generic/core/http/models/HttpResponse.java:56]: Does this need to be public ?

File: sdk/generic-sdk-core/generic-core-http-okhttp/src/main/java/com/generic/core/http/okhttp/implementation/OkHttpResponse.java [hunk 11]
```diff
@@ -4,34 +4,75 @@
...
+    }
+
+    /**
+     * Creates {@link com.generic.core.models.Headers Generic Core's headers} from {@link Headers OkHttp headers}.
+     *
+     * @param okHttpHeaders {@link Headers OkHttp headers}.
+     *
+     * @return {@link com.generic.core.models.Headers Generic Core's headers}.
+     */
+    static com.generic.core.models.Headers fromOkHttpHeaders(Headers okHttpHeaders) {
```
[the remark this answers is not in the document]
Comment (vcolin7) [reply sdk/generic-sdk-core/generic-core-http-okhttp/src/main/java/com/generic/core/http/okhttp/implementation/OkHttpResponse.java:48]: I checked the `generic-sdk-core-2` branch and this is still there in `[OkHttpResponse](https://github.com/Azure/azure-sdk-for-java/blob/c618cfa17f19e9845c50c30476952b6c1b957907/sdk/generic-sdk-core/generic-core-http-okhttp/src/main/java/com/generic/core/http/okhttp/implementation/OkHttpResponseBase.java#L51)`. If this was removed from Azure Core we can follow suit in a subsequent PR.

File: sdk/generic-sdk-core/generic-core/src/main/java/com/generic/core/http/models/HttpResponse.java (hunk 10 above)
Comment (vcolin7) [reply sdk/generic-sdk-core/generic-core/src/main/java/com/generic/core/http/models/HttpResponse.java:56]: No, I'm moving it to the implementation package.

File: sdk/generic-sdk-core/generic-core-http-okhttp/src/main/java/com/generic/core/http/okhttp/OkHttpHttpClient.java [hunk 12]
```diff
@@ -29,32 +27,32 @@
...
-
     final OkHttpClient httpClient;
 
     OkHttpHttpClient(OkHttpClient httpClient) {
         this.httpClient = httpClient;
     }
 
     @Override
-    public HttpResponse send(HttpRequest request) {
+    public com.generic.core.http.Response<?> send(HttpRequest request) {
```
Comment (JonathanGiles) [inline sdk/generic-sdk-core/generic-core-http-okhttp/src/main/java/com/generic/core/http/okhttp/OkHttpHttpClient.java:45]: Let's get rid of the full qualification of the type name.

File: sdk/generic-sdk-core/generic-core-http-okhttp/src/main/java/com/generic/core/http/okhttp/OkHttpHttpClient.java [hunk 13]
```diff
@@ -140,9 +138,10 @@ private static long getRequestContentLength(BinaryData content, Headers headers)
         return contentLength;
     }
 
-    private static HttpResponse toHttpResponse(HttpRequest request, okhttp3.Response response,
-                                               boolean eagerlyReadResponse, boolean ignoreResponseBody,
-                                               boolean eagerlyConvertHeaders) throws IOException {
+    private static com.generic.core.http.Response<?> toResponse(HttpRequest request, Response response,
```
Comment (JonathanGiles) [inline sdk/generic-sdk-core/generic-core-http-okhttp/src/main/java/com/generic/core/http/okhttp/OkHttpHttpClient.java:141]: Same here

File: sdk/generic-sdk-core/generic-core-http-okhttp/src/main/java/com/generic/core/http/okhttp/implementation/OkHttpResponse.java [hunk 14]
```diff
@@ -5,33 +5,110 @@
...
-public final class OkHttpResponse extends OkHttpResponseBase {
-    // Previously, this was 4096, but it is being changed to 8192 as that more closely aligns to what Netty uses as a
-    // default and will reduce the number of small allocations we'll need to make.
-    private static final int BYTE_BUFFER_CHUNK_SIZE = 8192;
+public class OkHttpResponse<T> implements com.generic.core.http.Response<T> {
+    private static final BinaryData EMPTY_BODY = BinaryData.fromBytes(new byte[0]);
 
+    private final BinaryData body;
+    private final Function<com.generic.core.http.Response<?>, ?> deserializationCallback;
+    private final com.generic.core.models.Headers headers;
```
Comment (JonathanGiles) [inline sdk/generic-sdk-core/generic-core-http-okhttp/src/main/java/com/generic/core/http/okhttp/implementation/OkHttpResponse.java:23]: same here (I'll stop mentioning it now - just get rid of all the full-qualifications)

File: sdk/generic-sdk-core/generic-core/src/main/java/com/generic/core/http/models/HttpRequest.java [hunk 15]
```diff
@@ -157,21 +160,21 @@ public BinaryData getBody() {
...
-     * @return This HttpRequest.
+     * @param content The {@link HttpRequest request} content.
+     *
+     * @return The updated {@link HttpRequest}.
      */
     public HttpRequest setBody(BinaryData content) {
         this.body = content;
 
-        // TODO (alzimmer): should the Content-Length header be removed if content is null?
+        // TODO (alzimmer): Should the Content-Length header be removed if the content is null?
```
Comment (JonathanGiles) [inline sdk/generic-sdk-core/generic-core/src/main/java/com/generic/core/http/models/HttpRequest.java:177]: Avoid making unrelated changes in your PR - it makes git blame and history more convoluted.

File: sdk/generic-sdk-core/generic-core/src/main/java/com/generic/core/implementation/http/RestProxy.java [hunk 16]
```diff
@@ -3,57 +3,58 @@
...
+ * Java objects as  * well as asynchronous single objects that resolve to a deserialized Java object.</p>
  */
 public final class RestProxy implements InvocationHandler {
     private final SwaggerInterfaceParser interfaceParser;
-    private final HttpPipeline httpPipeline;
     private final RestProxyImpl restProxyImpl;
 
     /**
-     * Create a RestProxy.
+     * Create a {@link RestProxy}.
```
Comment (JonathanGiles) [inline sdk/generic-sdk-core/generic-core/src/main/java/com/generic/core/implementation/http/RestProxy.java:30]: Focus on the impl - there is no point tweaking a bunch of JavaDoc for implementation code that won't ever have its javadoc generated. The end result is just a bunch of PR noise.

File: sdk/generic-sdk-core/generic-core/src/main/java/com/generic/core/http/policy/RetryPolicy.java [hunk 17]
```diff
@@ -148,14 +153,14 @@ default boolean shouldRetryCondition(RequestRetryCondition requestRetryCondition
         }
     }
 
-    private HttpResponse attempt(final HttpRequest httpRequest, final HttpPipelineNextPolicy next, final int tryCount,
-        final List<Exception> suppressed) {
+    private Response<?> attempt(final HttpRequest httpRequest, final HttpPipelineNextPolicy next, final int tryCount,
+                                final List<Exception> suppressed) throws IOException {
```
Comment (JonathanGiles) [inline sdk/generic-sdk-core/generic-core/src/main/java/com/generic/core/http/policy/RetryPolicy.java:157]: Why the IOException addition here?

File: sdk/generic-sdk-core/generic-core/src/main/java/com/generic/core/implementation/http/HttpResponse.java [hunk 18]
```diff
@@ -0,0 +1,133 @@
...
+import com.generic.core.models.BinaryData;
+import com.generic.core.models.Headers;
+
+import java.io.Closeable;
+import java.util.function.Function;
+
+/**
+ * The response of an {@link HttpRequest}.
+ */
+public class HttpResponse<T> implements Response<T>, Closeable {
```
Comment (JonathanGiles) [inline sdk/generic-sdk-core/generic-core/src/main/java/com/generic/core/implementation/http/HttpResponse.java:17]: I think this should remain public for now.

File: sdk/generic-sdk-core/generic-core/src/main/java/com/generic/core/http/models/HttpRequest.java [hunk 19]
```diff
@@ -216,20 +220,49 @@ public HttpRequest copy() {
...
+     * Set a {@link Function} that will handle deserialization for the body in the {@link Response} generated for this
+     * {@link HttpRequest request}.
+     *
+     * @param deserializationCallback The {@link Function} that will handle the {@link Response} body's deserialization.
+     *
+     * @return The updated {@link HttpRequest}.
+     */
+    public HttpRequest setResponseBodyDeserializationCallback(Function<Response<?>, ?> deserializationCallback) {
+        this.deserializationCallback = deserializationCallback;
+
```
Comment (JonathanGiles) [inline sdk/generic-sdk-core/generic-core/src/main/java/com/generic/core/http/models/HttpRequest.java:265]: This shouldn't be public API

File: sdk/generic-sdk-core/generic-core/src/main/java/com/generic/core/http/models/RequestOptions.java [hunk 20]
```diff
@@ -302,4 +306,28 @@ public RequestOptions setContext(Context context) {
...
+     * {@link HttpRequest request}.
+     *
+     * @param deserializationCallback The {@link Function} that will handle the {@link Response} body's deserialization.
+     *
+     * @return The updated {@link RequestOptions}.
+     */
+    public RequestOptions setResponseBodyDeserializationCallback(Function<Response<?>, ?> deserializationCallback) {
+        this.deserializationCallback = deserializationCallback;
+
+        return this;
```
Comment (JonathanGiles) [inline sdk/generic-sdk-core/generic-core/src/main/java/com/generic/core/http/models/RequestOptions.java:331]: This shouldn't be public API

**Review by JonathanGiles — CHANGES_REQUESTED** [review]

File: sdk/generic-sdk-core/generic-core-http-okhttp/src/main/java/com/generic/core/http/okhttp/implementation/OkHttpResponse.java (hunk 14 above)
Comment (vcolin7) [reply sdk/generic-sdk-core/generic-core-http-okhttp/src/main/java/com/generic/core/http/okhttp/implementation/OkHttpResponse.java:23]: We need it in some OkHttp classes as they also have a type named `Response`.

Comment (vcolin7) [reply sdk/generic-sdk-core/generic-core-http-okhttp/src/main/java/com/generic/core/http/okhttp/implementation/OkHttpResponse.java:23]: But I'll make sure to remove it from public method signatures.

File: sdk/generic-sdk-core/generic-core/src/main/java/com/generic/core/http/policy/RetryPolicy.java (hunk 17 above)
Comment (vcolin7) [reply sdk/generic-sdk-core/generic-core/src/main/java/com/generic/core/http/policy/RetryPolicy.java:157]: My bad, it got brought in from our "main" branch when I was solving some merge conflicts.

File: sdk/generic-sdk-core/generic-core-http-okhttp/src/main/java/com/generic/core/http/okhttp/implementation/OkHttpResponse.java (hunk 14 above)
Comment (vcolin7) [reply sdk/generic-sdk-core/generic-core-http-okhttp/src/main/java/com/generic/core/http/okhttp/implementation/OkHttpResponse.java:23]: I'll do you one better, let's remove fully qualified names for Generic Core types and just use them for OkHttp types when necessary (they're shorter).

File: sdk/generic-sdk-core/generic-core-http-okhttp/src/main/java/com/generic/core/http/okhttp/implementation/OkHttpResponse.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (samvaity) [reply sdk/generic-sdk-core/generic-core-http-okhttp/src/main/java/com/generic/core/http/okhttp/implementation/OkHttpResponse.java:60]: The copy paste removed the deprecation warning in the original file https://github.com/Azure/azure-sdk-for-java/blob/feature/generic-sdk-core-2/sdk/generic-sdk-core/generic-core-http-okhttp/src/main/java/com/generic/core/http/okhttp/implementation/OkHttpResponseBase.java#L50

We should not be carrying forward azure-core deprecated methods to generic-core. 
Could you add back the deprecation and do a follow up PR to remove it?

File: sdk/generic-sdk-core/generic-core-http-okhttp/src/main/java/com/generic/core/http/okhttp/implementation/OkHttpToGenericCoreHttpHeadersWrapper.java [hunk 21]
```diff
@@ -18,13 +18,13 @@
  * (as these headers are the result of an OkHttp response), we do so in any case, given the additional implementation
  * cost is minimal.
  */
-public final class OkHttpToAzureCoreHttpHeadersWrapper extends com.generic.core.models.Headers {
+public final class OkHttpToGenericCoreHttpHeadersWrapper extends com.generic.core.models.Headers {
```
Comment (samvaity) [inline sdk/generic-sdk-core/generic-core-http-okhttp/src/main/java/com/generic/core/http/okhttp/implementation/OkHttpToGenericCoreHttpHeadersWrapper.java:21]:
```suggestion
public final class OkHttpToCoreHttpHeadersWrapper extends com.generic.core.models.Headers {
```

File: sdk/generic-sdk-core/generic-core/src/main/java/com/generic/core/http/Response.java [hunk 22]
```diff
@@ -4,55 +4,87 @@
...
+     * Gets the {@link BinaryData} that represents the raw body of the response.
+     *
+     * @return The {@link BinaryData} containing the response body.
+     */
+    BinaryData getBody();
+
+    /**
+     * Closes the response content stream, if any.
+     */
+    default void close() throws IOException {
```
Comment (samvaity) [inline sdk/generic-sdk-core/generic-core/src/main/java/com/generic/core/http/Response.java:58]: why do we need this defaulted, leave it to the implenting classes to close the streams?

File: sdk/generic-sdk-core/generic-core-http-okhttp/src/main/java/com/generic/core/http/okhttp/implementation/OkHttpToGenericCoreHttpHeadersWrapper.java (hunk 21 above)
Comment (vcolin7) [reply sdk/generic-sdk-core/generic-core-http-okhttp/src/main/java/com/generic/core/http/okhttp/implementation/OkHttpToGenericCoreHttpHeadersWrapper.java]: I'll change this ina  separate PR, for now I've reverted unrelated changes.

File: sdk/generic-sdk-core/generic-core/src/main/java/com/generic/core/http/Response.java (hunk 22 above)
Comment (vcolin7) [reply sdk/generic-sdk-core/generic-core/src/main/java/com/generic/core/http/Response.java:58]: `HttpResponse` used to do it this way, I assume to not have all children be tasked with this. Only those that need it will override anyways.

Comment (vcolin7) [reply sdk/generic-sdk-core/generic-core/src/main/java/com/generic/core/http/Response.java]: We'll put this in `HttpResponse` for now.

File: sdk/generic-sdk-core/generic-core/src/main/java/com/generic/core/http/models/HttpResponse.java (hunk 10 above)
Comment (vcolin7) [reply sdk/generic-sdk-core/generic-core/src/main/java/com/generic/core/http/models/HttpResponse.java:51]: Sorry, I thought you meant the class, not the constructor. We can remove this constructor.

File: sdk/generic-sdk-core/generic-core-http-okhttp/src/main/java/com/generic/core/http/okhttp/implementation/OkHttpResponse.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (vcolin7) [reply sdk/generic-sdk-core/generic-core-http-okhttp/src/main/java/com/generic/core/http/okhttp/implementation/OkHttpResponse.java]: The deprecation suppression was there originally because [in Azure Core](https://github.com/Azure/azure-sdk-for-java/blob/e0a615cac870157c72b66b9a1b5b20cb6d78a153/sdk/core/azure-core-http-okhttp/src/main/java/com/azure/core/http/okhttp/implementation/OkHttpAsyncResponseBase.java#L84) the deprecated method [HttpHeaders.add()](https://github.com/Azure/azure-sdk-for-java/blob/e0a615cac870157c72b66b9a1b5b20cb6d78a153/sdk/core/azure-core/src/main/java/com/azure/core/http/HttpHeaders.java#L102) is used and I forgot to remove it when bringing in the class to Generic Core. We don't use the deprecated method in this version of Core so there's no need to suppress a warning that's not popping up.

File: sdk/generic-sdk-core/generic-core/src/main/java/com/generic/core/implementation/http/rest/RestProxyBase.java (hunk 6 above)
Comment (vcolin7) [reply sdk/generic-sdk-core/generic-core/src/main/java/com/generic/core/implementation/http/rest/RestProxyBase.java]: We're not explicitly casting anymore.

File: sdk/generic-sdk-core/generic-core-http-okhttp/src/main/java/com/generic/core/http/okhttp/implementation/OkHttpResponse.java [hunk 23]
```diff
@@ -4,34 +4,67 @@
...
-    public BinaryData getBody() {
-        return BinaryData.fromStream(this.responseBody.byteStream());
+    /**
+     * Creates {@link Headers Generic Core's headers} from {@link okhttp3.Headers OkHttp headers}.
+     *
+     * @param okHttpHeaders {@link okhttp3.Headers OkHttp headers}.
+     *
+     * @return {@link Headers Generic Core's headers}.
+     */
+    static Headers fromOkHttpHeaders(okhttp3.Headers okHttpHeaders) {
```
Comment (samvaity) [inline sdk/generic-sdk-core/generic-core-http-okhttp/src/main/java/com/generic/core/http/okhttp/implementation/OkHttpResponse.java:48]: Add back the deprecation warning. 
So that this can be done in a follow up PR.

Comment (vcolin7) [reply sdk/generic-sdk-core/generic-core-http-okhttp/src/main/java/com/generic/core/http/okhttp/implementation/OkHttpResponse.java:48]: We don't need it, we're not using a deprecated API.
